feat(terraform): add OCI protocol support for modules and providers

norman-zon opened the pull request:
Adds support for Terraform modules and providers hosted in OCI-compatible container registries using the `oci://` protocol.

Delegates OCI sources to Docker datasource for version resolution. Supports both version in separate field and embedded in URL.

Closes https://github.com/renovatebot/renovate/discussions/38927

<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->

## Context

Please select one of the below:

- [ x ] This closes an existing Issue: #38927

## AI assistance disclosure

<!-- We request this information to assist reviewers in identifying AI-generated errors and other issues specific to AI usage. While we typically permit the use of AI tools, we appreciate being notified when they are employed. -->

Did you use AI tools to create any part of this pull request?

Please select one option and, if yes, briefly describe how AI was used (e.g., code, tests, docs) and which tool(s) you used.

- [ x ] Yes — substantive assistance (AI generated non‑trivial portions of code, tests, or documentation): VSCode Copilot with Claude Sonnet 4.5 in Agent Mode


## Documentation (please check one with an [x])

- [ x ] I have updated the documentation

## How I've tested my work (please select one)

I have verified these changes via:

- [ x ] Newly added/modified unit tests


<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

norman-zon committed:
feat(terraform): add OCI protocol support for modules and providers

Adds support for Terraform modules and providers hosted in OCI-compatible
container registries using the `oci://` protocol.

Delegates OCI sources to Docker datasource for version resolution.
Supports both version in separate field and embedded in URL.

Closes https://github.com/renovatebot/renovate/discussions/38927

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/renovatebot/renovate?pullRequest=38934) <br/>All committers have signed the CLA.

siegenthalerroger commented:
pinging @zharinov and @viceice for reviews as they reviewed the last terraform items

viceice reviewed (commented):
> pinging @zharinov and @viceice for reviews as they reviewed the last terraform items

please fix merge conflicts first

secustor commented on lib/modules/manager/terraform/extractors/others/modules.ts:
```diff
@@ -67,8 +71,18 @@ export class ModuleExtractor extends DependencyExtractor {
+      const { registry, repository, tag } = ociRefMatch.groups;
+      dep.depName = `${registry}/${repository}`;
+      dep.packageName = `${registry}/${repository}`;
+      dep.registryUrls = [`https://${registry}`];
```
```suggestion
```
Not necessary

secustor commented on lib/modules/manager/terraform/extractors/others/modules.ts:
```diff
@@ -67,8 +71,18 @@ export class ModuleExtractor extends DependencyExtractor {
+      dep.datasource = DockerDatasource.id;
+      if (tag) {
+        dep.currentValue = tag;
+      }
```
This logic seems to be duplicated move this to a separate function

secustor commented on lib/modules/manager/terraform/extractors/others/modules.ts:
```diff
@@ -21,6 +22,9 @@ export const gitTagsRefMatchRegex = regEx(
 );
+export const ociRefMatchRegex = regEx(
+  /^oci:\/\/(?<registry>[^/:]+)\/(?<repository>[^:]+?)(?::(?<tag>.+))?$/,
+);
```
Seems to be duplicated

norman-zon committed:
refactor(terraform): address code review feedback for OCI support

cd - Move ociRefMatchRegex to shared location in base.ts
cd - Extract OCI parsing into parseOciRef function to reduce duplication
cd - Remove unnecessary registryUrls assignments
cd - Add skipReason when OCI tag is unspecified
cd - Inline oci.tf fixture to avoid triggering vulnerability scanners
cd - Refactor OCI regex tests to use it.each

viceice commented on lib/modules/manager/terraform/extract.spec.ts:
```diff
@@ -10,6 +10,53 @@ import { extractPackageFile } from './index.ts';
+    source = "oci://registry.example.com/terraform-modules/database:sha256:abc123"
+  }
+
+  module "traditional" {
```
no traditional cases needed for this test

norman-zon committed:
refactor(terraform): address PR review, round 2

- Replace custom ociRefMatchRegex/parseOciRef with getDep pattern,
reusing isOCIRegistry/removeOCIPrefix from helmv3/oci.ts
- Remove docs/development/oci-terraform-support.md
- Remove ociRefMatchRegex/parseOciRef from base.ts and modules.ts
- Simplify OCI test in extract.spec.ts
- Remove ociRefMatchRegex describe block from modules.spec.ts

norman-zon committed:
Merge branch 'main' into feat/terraform-oci

norman-zon committed:
Merge branch 'main' into feat/terraform-oci

norman-zon committed:
refactor(terraform): address PR review, round 3

extract duplicate function

Signed-off-by: Norman Stetter <85173861+norman-zon@users.noreply.github.com>

PatrickSpies commented:
@norman-zon: As per [opentofu-docs](https://opentofu.org/docs/language/modules/sources/#selecting-a-tag-or-digest) the version for modules from oci-registries needs to be a `tag`- or `digest`-query-parameter of the source-address:

```hcl
module "example" {
  source = "oci://example.com/repository-name?tag=v1.0.0"
}
```

Your tests currently cover unsupported formats

```hcl
module "example" {
  source = "oci://example.com/repository-name:1.2.3"
}
```
=>

> Error: Failed to download module
> 
>   on test.tf line 1:
>    1: module "example" {
> 
> Could not download module "example" (test.tf:1) source code from "oci://example.com/repository-name:1.2.3": error downloading 'oci://example.com/repository-name:1.2.3': invalid reference: invalid repository "example.com/repository-name:1.2.3"


```hcl
module "example" {
  source  = "oci://example.com/repository-name"
  version = "1.2.3"
}
```
=>

> Error: Invalid registry module source address
> 
>   on test.tf line 2, in module "example":
>    2:   source  = "oci://example.com/repository-name"
> 
> Failed to parse module registry address: a module registry source address must have either three or four slash-separated components.
> 
> OpenTofu assumed that you intended a module registry source address because you also set the argument "version", which applies only to registry modules.


Or do I miss something? Could you please check this? Thanks :)

norman-zon committed:
fix: only allow valid oci urls

oci sources only support query params for versioning

norman-zon committed:
feat: pass through registryaliases

norman-zon committed:
Merge branch 'main' into feat/terraform-oci

norman-zon committed:
fix: fix broken tests

norman-zon committed:
Merge branch 'main' into feat/terraform-oci

viceice commented on lib/modules/manager/terraform/util.ts:
```diff
@@ -49,6 +51,30 @@ export function getLockedVersion(
+): void {
+  // OCI sources only support versioning via `?tag=` or `?digest=` query params.
+  // See: https://opentofu.org/docs/language/modules/sources/#selecting-a-tag-or-digest
+  const [pathPart, queryPart] = removeOCIPrefix(source).split('?');
```
why not using `new URL`? is should parse because it looks like a valid url. it makes further parameters parsing much easier

norman-zon commented on lib/modules/manager/terraform/util.ts:
```diff
@@ -49,6 +51,30 @@ export function getLockedVersion(
+  const [pathPart, queryPart] = removeOCIPrefix(source).split('?');
+
+  // Strip optional `//subfolder` sub-path (e.g. `example.com/repo//modules/vpc`)
+  const imageRef = pathPart.replace(/\/\/.*$/, '');
```
Switched to regEx(/\/\/.+$/) with .+ as you suggested. A trailing // without actual subfolder content won't be stripped.

norman-zon committed:
fix: address PR review round 4

norman-zon committed:
Merge branch 'main' into feat/terraform-oci

norman-zon committed:
Merge branch 'main' into feat/terraform-oci

secustor commented on lib/modules/manager/terraform/util.ts:
```diff
@@ -49,6 +50,27 @@ export function getLockedVersion(
+  source: string,
+  registryAliases?: Record<string, string>,
+): void {
+  const url = new URL(source);
```
Use `parseUrl` form utils instead.

secustor reviewed (commented):
else LGTM

norman-zon committed:
fix: replace new URL with parseUrl

Signed-off-by: Norman Stetter <85173861+norman-zon@users.noreply.github.com>

norman-zon committed:
Merge branch 'main' into feat/terraform-oci

secustor reviewed (commented):
New path needs code coverage

norman-zon committed:
fix: add tests for new path

Signed-off-by: Norman Stetter <85173861+norman-zon@users.noreply.github.com>

norman-zon committed:
Merge branch 'main' into feat/terraform-oci

norman-zon commented:
Sorry for missing the coverage

secustor reviewed (approved):
LGTM, thanks for contributing!

norman-zon commented:
Thanks so much for your patience and support!
